runtime/src/main/java/com/dylibso/chicory/runtime/MStack.java

```diff
@@ -1,42 +1,797 @@
+     */
+    private Ref[] refs;
+
+    private int refDepth;
+
+    /**
+     * The stack protection base.
+     */
+    private int base;
+
+    /**
+     * Construct a new instance.
+     */
     public MStack() {
-        this.stack = new ArrayDeque<>();
+        this(256);
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param estimatedSize the estimated stack size (hint)
+     */
+    public MStack(int estimatedSize) {
+        estimatedSize = Math.max(4, estimatedSize);
+        types = new byte[estimatedSize];
+        words = new int[estimatedSize];
+        refs = new Ref[estimatedSize];
+    }
+
+    /**
+     * Get and set the stack protection base.
+     * Attempting to pop values below this base depth will result in an underflow exception.
+     * The base must be between 0 and the current stack depth (inclusive).
+     *
+     * @param newBase the new stack protection base
+     * @return the previous stack protection base
+     */
+    public int setProtectionBase(int newBase) {
+        if (newBase < 0 || newBase > depth) {
+            throw new IllegalArgumentException(
+                    "Stack protection base must be between 0 and the current depth");
+        }
+        try {
+            return base;
+        } finally {
+            base = newBase;
+        }
     }
 
     public void push(Value v) {
         if (v == null) throw new RuntimeException("Can't push null value onto stack");
-        this.stack.push(v);
+        switch (v.type()) {
+            case F64:
+                pushF64(v.asDouble());
+                break;
+            case F32:
+                pushF32(v.asFloat());
+                break;
+            case I64:
+                pushI64(v.asLong());
+                break;
+            case I32:
+                pushI32(v.asInt());
+                break;
+            case V128:
+                throw new UnsupportedOperationException();
+            case FuncRef:
+                if (v.asFuncRef() == Value.REF_NULL_VALUE) {
+                    pushFuncRef(null);
+                } else {
+                    // todo: this is not correct with respect to imports, host funcs, etc.
+                    pushFuncRef(new FuncRef(null, v.asFuncRef()));
+                }
+                break;
+            case ExternRef:
+                if (v.asExtRef() == Value.REF_NULL_VALUE) {
+                    pushExternRef(null);
+                } else {
+                    // todo: replace with proper mechanism
+                    pushExternRef(new ExternRef(CompatRefValue.values.get(v.asExtRef())));
+                }
+                break;
+            default:
+                // impossible
+                throw new IllegalStateException();
+        }
+    }
+
+    /**
+     * Push a zero value of the given type on to the stack.
+     *
+     * @param type the value type (must not be {@code null})
+     */
+    public void pushZero(ValueType type) {
```

Comment:
Probably I'm missing this, but I don't see where this method is used, neither I fully understand why we need a special method to push a zero on the stack instead of relying on other existing methods.

Reply:
This method would be used if/when stack frames are introduced: a zero value of each local type has to be pushed to reserve space for them. Since we start with a list of types, we can do it quickly this way e.g. `localTypes.forEach(stack::pushZero)`.